pilot/pkg/serviceregistry/kube/controller/filter/discoverynamespaces.go

```diff
@@ -142,6 +144,40 @@ func (d *discoveryNamespacesFilter) SelectorsChanged(
 	return
 }
 
+func (d *discoveryNamespacesFilter) SyncNamespaces() error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	newDiscoveryNamespaces := sets.NewString()
+
+	namespaceList, err := d.nsLister.List(labels.Everything())
+	if err != nil {
+		log.Errorf("error initializing discovery namespaces filter, failed to list namespaces: %v", err)
+		return err
+	}
+
+	// omitting discoverySelectors indicates discovering all namespaces
+	if len(d.discoverySelectors) == 0 {
+		for _, ns := range namespaceList {
+			newDiscoveryNamespaces.Insert(ns.Name)
+		}
+	}
+
+	// range over all namespaces to get discovery namespaces
+	for _, ns := range namespaceList {
+		for _, selector := range d.discoverySelectors {
+			if selector.Matches(labels.Set(ns.Labels)) {
+				newDiscoveryNamespaces.Insert(ns.Name)
+			}
+		}
+	}
+
+	// update filter state
+	d.discoveryNamespaces = newDiscoveryNamespaces
```

Comment:
can we deduplicate this code with SelectorsChanged? Seems like they are a copy?

Reply:
SelectorsChanged means discoveryNamespacesFilter's selectors have been changed, it is designed for mesh.watcher.selectors changed. During syncAll, we can use current discoveryNamespacesFilter's selectors, don't need to know where is `selectors`, although we can get from `c.opts.MeshWatcher.Mesh().DiscoverySelectors`